File: internal/models/event/event.go
package event

import (
	"time"
)

type EventType string

const (
	RegionalStage EventType = "REGIONAL_STAGE"
	Olympiad      EventType = "OLYMPIAD"
	Stage         EventType = "STAGE"
	ViewWorks     EventType = "VIEW_WORKS"
	Appeal        EventType = "APPEAL"
)

type Event struct {
	ID              uint      `gorm:"primaryKey"`
	Name            string    `gorm:"type:varchar(128);not null"`
	StartDate       time.Time `gorm:"not null"`
	EndDate         time.Time `gorm:"not null"`
	EventType       EventType `gorm:"type:event_type;not null"`
	PreviousEventID *uint     `gorm:"index"`
	PreviousEvent   *Event    `gorm:"foreignKey:PreviousEventID;constraint:OnUpdate:CASCADE,OnDelete:RESTRICT;"`
	Subject         string    `gorm:"type:varchar(128)"`
	// SubjectID       *uint            `gorm:"index,constraint:OnUpdate:CASCADE,OnDelete:RESTRICT;"`
	// Subject         *subject.Subject `gorm:"foreignKey:SubjectID"`
	AdditionalInfo string `gorm:"type:text"`
}

File: OlympiadReact/src/types/event.ts
export interface MyEvent {
    id?: number 
    name: string;
    start_date: Date;
    end_date: Date;
    previous_event_id?: number
    event_type: string;
    subject?: number;
    class_number?: number;
    additional_info?: string;
    events?: [MyEvent]
}

export const REGIONAL_STAGE = "REGIONAL_STAGE"
export const OLYMPIAD = "OLYMPIAD"
export const STAGE = "STAGE"
export const VIEW_WORKS = "VIEW_WORKS"
export const APPEAL = "APPEAL"
export const CLASS = "CLASS"


